core/src/main/scala/org/http4s/blaze/channel/ServerChannel.scala

```diff
@@ -1,67 +1,114 @@
-  final def join(): Unit = {
-    val l = new CountDownLatch(1)
+  /** Close the [[ServerChannel]] and execute any shutdown hooks
+    *
+    * @note this method is idempotent
+    */
+  final def close(): Unit = {
+    val run = shutdownHooks.synchronized {
+      if (state != Open) false
+      else {
+        state = Closing
+        // Need to close the channel, get all the hooks, then notify the listeners
+        closeChannel()
+        true
+      }
+    }
 
-    if (!addShutdownHook(() => l.countDown()))
-      l.countDown()
+    // We move this outside of the lock to ensure that we don't schedule work
+    // inside the lock since the EC could run the task immediately.
+    if (run) scheduleHook()
+  }
 
-    l.await()
+  /** Wait for this server channel to close, including execution of all successfully
+    * registered shutdown hooks.
+    */
+  final def join(): Unit = shutdownHooks.synchronized {
+    while (state != Closed) {
+      shutdownHooks.wait()
+    }
   }
 
   /** Add code to be executed when the [[ServerChannel]] is closed
     *
     * @param f hook to execute on shutdown
-    * @return true if the hook was successfully registered, false otherwise
+    * @return true if the hook was successfully registered, false otherwise.
     */
-  final def addShutdownHook(f: () => Unit): Boolean = {
-    @tailrec
-    def go(): Boolean = shutdownHooks.get() match {
-      case null                                                     => false
-      case hooks if(shutdownHooks.compareAndSet(hooks, hooks :+ f)) => true
-      case _                                                        => go()
+  final def addShutdownHook(f: () => Unit)(implicit ec: ExecutionContext = Execution.directec): Boolean = {
+    shutdownHooks.synchronized {
+      if (state != Open) false
+      else {
+        shutdownHooks += Hook(f, ec)
+        true
+      }
     }
+  }
 
-    go()
+  // Checks if we have a hook, and if we do, schedule it and then check again
+  private[this] def scheduleHook(): Unit = {
+    val hook = shutdownHooks.synchronized {
+      if (shutdownHooks.nonEmpty) Some(shutdownHooks.dequeue())
+      else {
+        // all our hooks are done! Switch to closed state, notify
+        // all the listeners
+        state = Closed
+        shutdownHooks.notifyAll()
+        None
+      }
+    }
+    hook match {
+      case Some(hook) => hook.ec.execute(new HookRunnable(hook.task))
+      case None => ()
+    }
   }
 
-  private def runShutdownHooks(): Unit = {
-    val hooks = shutdownHooks.getAndSet(null)
-    if (hooks != null) {
-      hooks.foreach { f =>
-        try f()
-        catch { case NonFatal(t) => logger.error(t)(s"Exception occurred during Channel shutdown.") }
+  private case class Hook(task: () => Unit, ec: ExecutionContext)
+
+  // Bundles the task of executing a hook and scheduling the next hook
+  private[this] class HookRunnable(hook: () => Unit) extends Runnable {
+    override def run(): Unit = {
+      try hook()
```

Comment:
What guarantees to we have about `hook` here. 

If a hook is created in such a way it is put back in queue, this could create an infinite loop. As The `HookRunnable` is run outside of the synchronized lock for obvious lock contention reasons.